### ISSUE TITLE
Undefined atom types referenced in SMARTS definition from xml writer

### ISSUE THREAD

**Issue by rmatsum836** [mosdef-hub/foyer#255]
**Bug summary**
Testing #254, I tried reading in the xml file that was output from `write_foyer`.  
When attempting to create a `Forcefield` object from this xml, an error occurs due to undefined atom types being referenced in the SMARTS definitions.

Based on this error, we should parse out the atom types that don't exist in the xml from the SMARTS definitions. 

**Code to reproduce the behavior**

```python
lopes = Forcefield('co-test.xml')
foyer.exceptions.ValidationError: Reference to undefined atomtype 'lopes_016' in SMARTS string 'H[C;%lopes_016,%lopes_017,%lopes_018,%lopes_019]' at line 12
```
The issue is that `lopes_016` isn't an atom type in this specific system, so it isn't written out to the xml from `write_foyer`.  

**Software versions**

- Which version of Foyer are you using? 0.6.2
- Which version of Python (`python --version`)? 3.7.3
- Which operating system? Mac OS 10.14.4

**Comment by daico007** [comment]
@rmatsum836 Can you upload the `co-test.xml` file?

**Comment by rmatsum836** [comment]
https://gist.github.com/rmatsum836/568a1c7218f127542ac8e8fd3c6b485a

**Comment by ahy3nz** [comment]
I think there are 3 options:
1) Turn off the `Validator` just for the sake of testing this

2) If an atomtype is referenced in a smarts description, then include that atomtype in the ffxml. 

2a) Include the extra atomtype, but not the smarts string. This extra atomtype might not need a sophisticated smarts string if this atomtype is never even used in a particular chemical system. Really this option would just get around any complaints from the `Validator`. While the extra atomtype would be present, it might never get used?

2b) Include the extra atomtype, but include the smarts string. This would be a more "complete" ffxml, but then you'd have to add even more atomtypes depending on the extra atomtype's smarts string

3) Modify the smarts string to trim out any atomtypes that aren't present in the chemical system? Not sure if this will incur any consequences if we modify the smarts definitions so the minimal-FFXML is purely self-described

**Comment by rmatsum836** [comment]
Option 3: modifying the smarts string to to trim out any non-existent atom types was the first approach I tried.  It's not immediately clear to me how to parse the string in a general enough manner, so I might go with 2b.

**Comment by daico007** [comment]
This seems like the block of code that shows how the atomtype is picked out in the validator. I think we can use the same logic to single out and remove non-existent atom types. I am just not sure how it gonna affect down the line though. 
```
            # make sure referenced labels exist
            smarts_graph = SMARTSGraph(smarts_string, parser=self.smarts_parser,
                                       name=name, overrides=entry.attrib.get('overrides'))
            for atom_expr in nx.get_node_attributes(smarts_graph, name='atom').values():
                labels = atom_expr.find_data('has_label')
                for label in labels:
                    atom_type = label.children[0][1:]
                    if atom_type not in self.atom_type_names:
                        undefined = ValidationError(
                            "Reference to undefined atomtype '{}' in SMARTS "
                            "string '{}' at line {}".format(atom_type, entry.attrib['def'], entry.sourceline),
                            None, entry.sourceline)
                        errors.append(undefined)
```

**Comment by rmatsum836** [comment]
Thanks I'll take a look.

**Comment by uppittu11** [comment]
Fixed with #256
